src/System/TimedScopes/Correlation.cs

```diff
@@ -0,0 +1,302 @@
+	/// Correlation data by default only applies to the current thread and is lost when a thread hop
+	/// occurs (be it ASP.NET underlying thread hop or an explicit thread change). By calling
+	/// CorrelationStart with cloned CorrelationData, the correlation can be transferred to the new
+	/// thread.
+	/// </remarks>
+	public sealed class Correlation
+	{
+		/// <summary>
+		/// Constructor
+		/// </summary>
+		/// <param name="correlationHandler">correlation handler, used to store correlation data in underlying logging system</param>
+		/// <param name="callContextManager">call context manager, to handle context operations</param>
+		/// <param name="machineInformation">machine information</param>
+		public Correlation(ICorrelationStorage correlationHandler, ICallContextManager callContextManager, IMachineInformation machineInformation)
+		{
+			CorrelationHandler = correlationHandler;
+			CallContextManager = callContextManager;
+			MachineInformation = machineInformation;
+		}
+
+
+		/// <summary>
+		/// Current call context
+		/// </summary>
+		private static ICallContext CallContext => CallContextManager.CallContextHandler(MachineInformation);
+
+
+		/// <summary>
+		/// Event raised when the correlation has started
+		/// </summary>
+		public event EventHandler<CorrelationEventArgs> CorrelationStarted;
+
+
+		/// <summary>
+		/// Event raised when correlation data has been added
+		/// </summary>
+		public event EventHandler<CorrelationEventArgs> CorrelationDataAdded;
+
+
+		/// <summary>
+		/// Event raised when the correlation has ended
+		/// </summary>
+		public event EventHandler<CorrelationEventArgs> CorrelationEnded;
+
+
+		/// <summary>
+		/// Key used to store the correlation context in the current HttpContext
+		/// </summary>
+		private const string CorrelationContextKey = "ULSLogging.correlations";
+
+
+		/// <summary>
+		/// Key used to store if direct logging should be used
+		/// </summary>
+		private const string LoggingContextKey = "ULSLogging.directLogging";
+
+
+		/// <summary>
+		/// Get the current correlation if set
+		/// </summary>
+		/// <exception cref="InvalidCastException">Data exists on the CallContext
+		/// but it isn't a valid CorrelationData object.</exception>
+		public CorrelationData CurrentCorrelation
+		{
+			get
+			{
+				return GetCorrelationData();
+			}
+			private set
+			{
+				ICallContext existingContext = CallContext.ExistingCallContext();
+				if (existingContext != null)
+				{
+					if (value != null)
+					{
+						existingContext.Data[CorrelationContextKey] = value;
+					}
+					else
+					{
+						existingContext.Data.Remove(CorrelationContextKey);
+					}
+				}
+			}
+		}
+
+
+		/// <summary>
+		/// Should log directly to the underlying log handler
+		/// </summary>
+		public bool ShouldLogDirectly
+		{
+			get
+			{
+				CorrelationData correlation = CurrentCorrelation;
+				if (correlation != null && correlation.ShouldLogDirectly)
```

Comment:
>correlation != null && correlation.ShouldLogDirectly [](start = 8, length = 52)

Will correlation?.ShouldLogDirectory work in this context?

Reply:
yes, I changed it.